### ISSUE TITLE
Decouple adblock engines from component loading methods

### ISSUE THREAD

**Issue by antonok-edm** [brave/brave-browser#19366]
Initially, adblock engines were only loaded with sources from Brave's component server. Over time, additional adblock engines were added that load from the browser's local preferences (custom filters) or from remote 3rd party servers (subscription lists). The code is still structured to support the initial use-case, which is confusing and limits future improvements. Once this is implemented properly, it will unblock future work like being able to combine the engines together for performance reasons, or untangling threading concerns.

This should ultimately be achieved by removing `BraveComponent` from the `AdBlockService` inheritance tree, and replacing it with a common `SourceProvider` interface that is generic enough to be implemented by the necessary components, custom filters loader, or subscription lists.

**Comment by LaurenWags** [comment]
@antonok-edm could you help @brave/legacy_qa with specific test cases (particularly in high impact areas) that we could use to check this one? There are a number of previous issues related to adblock, so any assistance you can provide in focusing in on the scope would be appreciated. cc @rebron due to this being a large refactor per https://github.com/brave/brave-core/pull/10994#issue-1050464565

**Comment by stephendonner** [comment]
Verified `PASSED` using 

Brave | 1.37.92 Chromium: 99.0.4844.51 (Official Build) beta (x86_64)
-- | --
Revision | d537ec02474b5afe23684e7963d538896c63ac77-refs/branch-heads/4844@{#875}
OS | macOS Version 11.6.3 (Build 20G415)

Testcase 1: Verify when Brave flag is enabled, cookie notice is dismissed and when disabled, cookie notice is displayed again

I. Enable `#brave-adblock-cookie-list-default` in `brave://flags`
  - Install 1.37.x
  - Launch `Brave`
  - Visited https://digikey.com  
  - `Cookie notice` displayed on the top
  - Visit `brave://flags` and enabled `#brave-adblock-cookie-list-default` 
  - Clicked `Relaunch` to relaunch the browser
  - Visited https://digikey.com and verified `cookie notice` no longer is displayed
  - Visited `brave://adblock` and verified an entry for `Easylist-Cookie List - Filter Obtrusive Cookie Notices` is available under 
  `Additional filters` and is checked

II. Disable #brave-adblock-cookie-list-default in brave://flags 

  - Visited `brave://flag` and disabled `#brave-adblock-cookie-list-default` 
  - Clicked `Relaunch` to relaunch the browser
  - Visited https://digikey.com and verified `cookie notice` displayed again
  - Verified `Easylist-Cookie List - Filter Obtrusive Cookie Notices` is unchecked

1|2|3|4|5|6|7
----|----|----|----|----|----|----
<img width="1312" alt="Screen Shot 2022-03-16 at 1 27 29 PM" src="https://user-images.githubusercontent.com/387249/158685310-2ccf8267-6682-46d3-bc7f-6644ee71c029.png"> | <img width="1312" alt="Screen Shot 2022-03-16 at 1 27 20 PM" src="https://user-images.githubusercontent.com/387249/158685366-613d1db2-216e-4548-818a-5fa2920448fc.png"> | <img width="1312" alt="Screen Shot 2022-03-16 at 1 26 12 PM" src="https://user-images.githubusercontent.com/387249/158685434-efc9d82a-16d9-40ea-b364-0c4fb159507f.png"> | <img width="1312" alt="Screen Shot 2022-03-16 at 1 25 25 PM" src="https://user-images.githubusercontent.com/387249/158685993-66709d82-a0a8-4c19-a8c4-82ca93e31522.png"> | <img width="1312" alt="Screen Shot 2022-03-16 at 1 22 51 PM" src="https://user-images.githubusercontent.com/387249/158685818-e2a66910-8594-42f3-b31e-56396cb3506c.png"> | <img width="1312" alt="Screen Shot 2022-03-16 at 1 20 05 PM" src="https://user-images.githubusercontent.com/387249/158686197-9fda624b-3919-48f4-b1c5-67e4585b7c35.png"> |<img width="1312" alt="Screen Shot 2022-03-16 at 1 20 24 PM" src="https://user-images.githubusercontent.com/387249/158685855-d0c97ed1-e509-46f6-b947-1b087985cb5d.png">



Testcase 2 - Verify cookie banner is dismissed with aggressive mode in Shields

I. Enable #brave-adblock-cookie-list-default in brave://flags 

  - Clean profile
  - Visited https://gulfnews.com  
  - Cookie notice displayed at the bottom
  - Visited `brave://flags` and enabled `#brave-adblock-cookie-list-default` 
  - Clicked `Relaunch` to relaunch the browser
  - Visited https://gulfnews.com but the `cookie notice` still displayed
  - Visited brave://adblock and verified an entry for `Easylist-Cookie List - Filter Obtrusive Cookie Notices` is available and is checked
  - Click on the `Shields` to select to `Aggressive` mode in `Trackers & ads blocked``
  - Verified `cookie notice` no longer displayed

II. Disable `#brave-adblock-cookie-list-default` in `brave://flags`

  - Visited `brave://flag` and disabled `#brave-adblock-cookie-list-default` 
  - Clicked `Relaunch` to relaunch the browser
  - Visited https://gulfnews.com and verified `cookie notice` displayed at the bottom
  - Verified `Easylist-Cookie List - Filter Obtrusive Cookie Notices` in `brave://adblock` is unchecked

1|2|3|4|5|6|7
---|---|---|---|---|---|---
<img width="1246" alt="Screen Shot 2022-03-16 at 1 55 07 PM" src="https://user-images.githubusercontent.com/387249/158689484-da161cca-97b3-4e6f-b256-e3d203724234.png"> | <img width="1246" alt="Screen Shot 2022-03-16 at 1 55 35 PM" src="https://user-images.githubusercontent.com/387249/158689867-ff462963-aa25-4aba-8732-4107cd375210.png"> | <img width="1246" alt="Screen Shot 2022-03-16 at 1 56 04 PM" src="https://user-images.githubusercontent.com/387249/158689961-2b37d64e-56f9-4bfc-983a-c8d9772d9264.png"> | <img width="1246" alt="Screen Shot 2022-03-16 at 1 57 09 PM" src="https://user-images.githubusercontent.com/387249/158689984-52a4fc1b-835b-4aca-9a1b-7aaedf08cbed.png"> | <img width="1312" alt="Screen Shot 2022-03-16 at 1 52 46 PM" src="https://user-images.githubusercontent.com/387249/158689046-87d399a7-3014-4595-a3ef-ab6c5e101a8c.png"> | <img width="1312" alt="Screen Shot 2022-03-16 at 1 43 42 PM" src="https://user-images.githubusercontent.com/387249/158688236-003e0ac9-8b7a-48f3-9b37-9f627b1da2cb.png"> | <img width="1246" alt="Screen Shot 2022-03-16 at 2 02 11 PM" src="https://user-images.githubusercontent.com/387249/158690353-c12b742f-ad3c-4da2-bbad-16dc7e157542.png">




Testcase 3 - Manually toggle `Easylist-Cookie List - Filter Obtrusive Cookie Notices' 

- Clean profile
- Visited `brave://adblock` and checked `Easylist-Cookie List - Filter Obtrusive Cookie Notices`  
- Visited `brave://flag` and enabled `#brave-adblock-cookie-list-default` 
- Clicked `Relaunch` to relaunch the browser

NOTE: after enabling `Easylist-Cookie List` via `brave://adblock`, my experience diverges from @MadhaviSeelam in that I no longer see the cookie dialog here.

- Visited `brave://adblock` and verified `Easylist-Cookie List - Filter Obtrusive Cookie Notices` still checked
- Returned to `brave://flags` disabled #brave-adblock-cookie-list-default flag and restarted the browser.
- Visited `brave://adblock` and verified an entry for `Easylist-Cookie List - Filter Obtrusive Cookie Notices` is still checked
- Visited `brave://adblock` and unchecked `Easylist-Cookie List - Filter Obtrusive Cookie Notices`  
- Visited `brave://flags` and enabled `#brave-adblock-cookie-list-default` flag and restarted the browser.
- Visited https://digikey.com and confirmed `cookie notice` is displayed

1|2|3|4|5|6|7
---|---|---|---|---|---|---
<img width="1110" alt="Screen Shot 2022-03-16 at 2 05 33 PM" src="https://user-images.githubusercontent.com/387249/158691334-e894cf4b-9cae-4e5d-9d9e-dc1fdfe0667b.png"> | <img width="1110" alt="Screen Shot 2022-03-16 at 2 05 56 PM" src="https://user-images.githubusercontent.com/387249/158691379-1c5bb5b0-320b-4825-a893-81b65926732f.png"> | <img width="1636" alt="Screen Shot 2022-03-16 at 2 06 22 PM" src="https://user-images.githubusercontent.com/387249/158691925-6beededd-93a1-487c-83b2-c5e6c097ec38.png"> | <img width="1636" alt="Screen Shot 2022-03-16 at 2 14 29 PM" src="https://user-images.githubusercontent.com/387249/158692206-0667fe57-e627-4bb0-9f2a-e64734ea150d.png"> | <img width="1636" alt="Screen Shot 2022-03-16 at 2 07 49 PM" src="https://user-images.githubusercontent.com/387249/158691486-67dadf48-7301-4d7e-86b4-78fc05fe9859.png"> | <img width="1636" alt="Screen Shot 2022-03-16 at 2 07 49 PM" src="https://user-images.githubusercontent.com/387249/158692356-ef3164da-db77-4ca7-9bf2-0fa9a92e2027.png"> | <img width="1636" alt="Screen Shot 2022-03-16 at 2 07 08 PM" src="https://user-images.githubusercontent.com/387249/158692347-98a13d6f-64c0-4f0c-b8ba-4d46a1bf8889.png"> | <img width="1636" alt="Screen Shot 2022-03-16 at 2 07 58 PM" src="https://user-images.githubusercontent.com/387249/158692730-821d6f1a-2f45-4cd1-91bc-ebc85d84f0e5.png">



Note: In addition to above testing, verified `Shields` functionality works same as expected by comparing `1.37.x` with`1.36.x `.

**Comment by Uni-verse** [comment]
Verification in Progress (Android) 

Test Case 1: 
Test Case 2: 
Test Case 3:

**Comment by btlechowski** [comment]
Verification passed on

Brave | 1.37.101 Chromium: 99.0.4844.83 (Official Build) beta (64-bit)
-- | --
Revision |  b11086e62d7c1a44b0942ac5568d22a425c7ae35-refs/branch-heads/4844_74@{#5}
OS | Ubuntu 20 LTS

Testcase 1: Verify when Brave flag is enabled, cookie notice is dismissed and when disabled, cookie notice is displayed again

I. Enable `#brave-adblock-cookie-list-default` in `brave://flags`
  - Install 1.37.x
  - Launch `Brave`
  - Visited https://digikey.com
  - `Cookie notice` displayed on the top
  - Visit `brave://flags` and enabled `#brave-adblock-cookie-list-default` 
  - Clicked `Relaunch` to relaunch the browser
  - Visited https://digikey.com and verified `cookie notice` no longer is displayed
  - Visited `brave://adblock` and verified an entry for `Easylist-Cookie List - Filter Obtrusive Cookie Notices` is available under 
  `Additional filters` and is checked

II. Disable #brave-adblock-cookie-list-default in brave://flags 

  - Visited `brave://flag` and disabled `#brave-adblock-cookie-list-default` 
  - Clicked `Relaunch` to relaunch the browser
  - Visited https://digikey.com and verified `cookie notice` displayed again
  - Verified `Easylist-Cookie List - Filter Obtrusive Cookie Notices` is unchecked

<details>

![image](https://user-images.githubusercontent.com/34715963/160205060-a278b2bd-7e56-42cc-aeba-efc201e1436c.png)
![image](https://user-images.githubusercontent.com/34715963/160206369-529b954b-24a1-4761-ae45-f1ffdebd9aa3.png)

![image](https://user-images.githubusercontent.com/34715963/160206266-d61e093a-efd3-468f-a8e7-6c59b505e15d.png)
![image](https://user-images.githubusercontent.com/34715963/160206526-ff3d379a-34fc-4668-b825-3d464fdf466b.png)




</details>



Testcase 2 - Verify cookie banner is dismissed with aggressive mode in Shields

I. Enable #brave-adblock-cookie-list-default in brave://flags 

  - Clean profile
  - Visited https://gulfnews.com  
  - Cookie notice displayed at the bottom
  - Visited `brave://flags` and enabled `#brave-adblock-cookie-list-default` 
  - Clicked `Relaunch` to relaunch the browser
  - Visited https://gulfnews.com but the `cookie notice` still displayed
  - Visited brave://adblock and verified an entry for `Easylist-Cookie List - Filter Obtrusive Cookie Notices` is available and is checked
  - Click on the `Shields` to select to `Aggressive` mode in `Trackers & ads blocked``
  - Verified `cookie notice` no longer displayed

II. Disable `#brave-adblock-cookie-list-default` in `brave://flags`

  - Visited `brave://flag` and disabled `#brave-adblock-cookie-list-default` 
  - Clicked `Relaunch` to relaunch the browser
  - Visited https://gulfnews.com and verified `cookie notice` displayed at the bottom
  - Verified `Easylist-Cookie List - Filter Obtrusive Cookie Notices` in `brave://adblock` is unchecked

<details>

![image](https://user-images.githubusercontent.com/34715963/160206960-d9ddc400-658a-4879-9576-60d65786fc00.png)
![image](https://user-images.githubusercontent.com/34715963/160207161-6dc491cf-642a-49ff-a5d5-77690fb07dfd.png)
![image](https://user-images.githubusercontent.com/34715963/160207125-d1180004-14a2-4af6-b67e-365f2d18eace.png)
![image](https://user-images.githubusercontent.com/34715963/160207190-866ba868-7349-4b3d-ae9d-2314ed298264.png)
![image](https://user-images.githubusercontent.com/34715963/160207235-af448532-1ad5-48aa-b9dc-3d5c360319ca.png)



</details>


Testcase 3 - Manually toggle `Easylist-Cookie List - Filter Obtrusive Cookie Notices' 

- Clean profile
- Visited `brave://adblock` and checked `Easylist-Cookie List - Filter Obtrusive Cookie Notices`  
- Visited `brave://flag` and enabled `#brave-adblock-cookie-list-default` 
- Clicked `Relaunch` to relaunch the browser

NOTE: after enabling `Easylist-Cookie List` via `brave://adblock`, my experience diverges from @MadhaviSeelam in that I no longer see the cookie dialog here.

- Visited `brave://adblock` and verified `Easylist-Cookie List - Filter Obtrusive Cookie Notices` still checked
- Returned to `brave://flags` disabled #brave-adblock-cookie-list-default flag and restarted the browser.
- Visited `brave://adblock` and verified an entry for `Easylist-Cookie List - Filter Obtrusive Cookie Notices` is still checked
- Visited `brave://adblock` and unchecked `Easylist-Cookie List - Filter Obtrusive Cookie Notices`  
- Visited `brave://flags` and enabled `#brave-adblock-cookie-list-default` flag and restarted the browser.
- Visited https://digikey.com and confirmed `cookie notice` is not displayed

<details>

![image](https://user-images.githubusercontent.com/34715963/160207532-8153585f-ac95-4bfa-909e-8392ddf6dec0.png)
![image](https://user-images.githubusercontent.com/34715963/160207621-8732a16f-c1e0-4fea-a92d-8035f2941be2.png)


</details>




Note: In addition to above testing, verified `Shields` functionality works same as expected by comparing `1.37.x` with`1.36.x ` on several of the top Alexa 20 sites, and other favorites of mine.
